rust-runtime/aws-smithy-client/src/quota.rs

```diff
@@ -0,0 +1,156 @@
+        Self {
+            inner: Arc::new(Mutex::new(INITIAL_RETRY_TOKENS)),
+            max_tokens: INITIAL_RETRY_TOKENS,
+            retry_cost: RETRY_COST,
+            default_release_amount: 1,
+            timeout_retry_cost: 10,
+        }
+    }
+}
+
+impl QuotaState {
+    /// Create a new `QuotaState`.
+    pub fn new(
+        initial_quota: usize,
+        retry_cost: usize,
+        default_release_amount: usize,
+        timeout_retry_cost: usize,
+    ) -> Self {
+        Self {
+            inner: Arc::new(Mutex::new(initial_quota)),
+            max_tokens: initial_quota,
+            retry_cost,
+            default_release_amount,
+            timeout_retry_cost,
+        }
+    }
+
+    /// Return the number of tokens available in this token bucket.
+    pub fn available(&self) -> usize {
+        // TODO don't unwrap here
+        *self.inner.lock().unwrap()
+    }
+
+    /// Release some number of tokens. If no number is passed, release tokens equal to the
+    /// `default_release_amount` passed when creating this `QuotaState`. No matter how many
+    /// tokens are released, the total amount of tokens in the bucket will be capped to the
+    /// initial amount of tokens set when creating this `QuotaState`.
+    pub fn release(&self, value: Option<usize>) {
+        let mut quota = self.inner.lock().unwrap();
+        *quota = (*quota + value.unwrap_or(self.default_release_amount)).min(self.max_tokens);
+    }
+
+    /// Attempt to acquire retry quota for `ErrorKind`
+    ///
+    /// If quota is available, the amount of quota consumed is returned
+    /// If no quota is available, `None` is returned.
+    pub fn acquire(&self, err: &ErrorKind) -> Option<usize> {
+        let retry_cost = if err == &ErrorKind::TransientError {
+            self.timeout_retry_cost
+        } else {
+            self.retry_cost
+        };
+
+        let mut quota = self.inner.lock().unwrap();
+        if retry_cost > *quota {
+            None
+        } else {
+            *quota -= retry_cost;
+            Some(retry_cost)
+        }
+    }
+}
+
+/// A service that wraps another service, adding the ability to set a quota for requests
+/// handled by the inner service.
+#[derive(Clone, Debug)]
+pub struct QuotaService<S> {
+    inner: S,
+    state: QuotaState,
+}
+
+impl<S> QuotaService<S> {
+    /// Create a new `QuotaService`
+    pub fn new(inner: S, state: QuotaState) -> Self {
+        Self { inner, state }
+    }
+}
+
+impl<H, R, InnerService, E> tower::Service<Operation<H, R>> for QuotaService<InnerService>
+where
+    InnerService: tower::Service<Operation<H, R>, Error = SdkError<E>>,
+{
+    type Response = InnerService::Response;
+    type Error = SdkError<E>;
+    type Future = InnerService::Future;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        if self.state.available() == 0 {
+            // If no quota is available, report pending
+            Poll::Pending
```

Comment:
this won't quite work—we need to schedule a wakeup when state goes back above 0. As written, if state goes to 0, that task will never get woken up